File: src/mnemonic/logger.py
"""進捗表示およびログ出力のインターフェース定義

このモジュールは、Mnemonicのビルド進捗表示とログ出力のためのインターフェースを定義する。
VerboseLevel (詳細ログレベル)に応じた出力制御を行い、
CLIでのビルド進捗をユーザーにわかりやすく表示するために使用される。
"""

from __future__ import annotations

import re
import sys
from dataclasses import dataclass
from datetime import datetime
from enum import IntEnum
from pathlib import Path
from typing import TYPE_CHECKING, Any, Protocol, TextIO

if TYPE_CHECKING:
    from mnemonic.pipeline import PipelinePhase


class VerboseLevel(IntEnum):
    """詳細ログレベル

    ログ出力の詳細度を制御するための列挙型。
    QUIET: エラーのみ出力
    NORMAL: 進捗バーとサマリ出力
    VERBOSE: 変換ファイル一覧も出力（-vオプション）
    DEBUG: 外部コマンド実行ログも出力（-vvオプション）
    """

    QUIET = -1
    NORMAL = 0
    VERBOSE = 1
    DEBUG = 2


class ProgressDisplay(Protocol):
    """進捗表示のプロトコル

    ビルドパイプラインの各フェーズの進捗を表示するためのインターフェース。
    CLI UIやログ出力などで進捗を可視化する際に使用する。
    """

    def start(self, phase: PipelinePhase, total: int) -> None:
        """フェーズ開始を表示する

        Args:
            phase: 開始するパイプラインフェーズ
            total: 処理対象の総数
        """
        ...

    def update(self, current: int, message: str = "") -> None:
        """進捗を更新する

        Args:
            current: 現在の進捗（処理済みアイテム数）
            message: 追加の進捗メッセージ（オプション）
        """
        ...

    def finish(self, success: bool, message: str = "") -> None:
        """フェーズ終了を表示する

        Args:
            success: フェーズが成功したか
            message: 終了メッセージ（オプション）
        """
        ...


@dataclass
class LogConfig:
    """ログ設定

    ログ出力の動作を制御するための設定データクラス。
    詳細レベル、ファイル出力、色やemojiの使用有無を設定できる。

    Attributes:
        verbose_level: ログの詳細度レベル
        log_file: ログ出力先ファイルパス（Noneの場合はファイル出力なし）
        use_color: カラー出力を使用するか
        use_emoji: emoji表示を使用するか
    """

    verbose_level: VerboseLevel = VerboseLevel.NORMAL
    log_file: Path | None = None
    use_color: bool = True
    use_emoji: bool = True


class BuildLogger:
    """ビルドログ出力クラス

    ビルドパイプラインのログ出力を管理するクラス。
    VerboseLevelに応じてメッセージのフィルタリングを行い、
    進捗表示インスタンスの作成も担当する。

    使用例:
        >>> config = LogConfig(verbose_level=VerboseLevel.VERBOSE)
        >>> logger = BuildLogger(config)
        >>> logger.info("ビルドを開始します")
        >>> logger.verbose("game.exe を処理中")
    """

    _ANSI_ESCAPE_PATTERN = re.compile(r"\x1b\[[0-9;]*m")

    def __init__(self, config: LogConfig) -> None:
        """ロガーを初期化する

        Args:
            config: ログ設定
        """
        self._config = config
        self._log_file: TextIO | None = None
        if config.log_file:
            # クラス自体がコンテキストマネージャとして動作し、__exit__でファイルを閉じる
            self._log_file = open(config.log_file, "w", encoding="utf-8")  # noqa: SIM115

    def __enter__(self) -> BuildLogger:
        """コンテキストマネージャのエントリポイント"""
        return self

    def __exit__(self, *args: object) -> None:
        """コンテキストマネージャの終了処理"""
        if self._log_file:
            self._log_file.close()
            self._log_file = None

    @property
    def config(self) -> LogConfig:
        """ログ設定を取得する

        Returns:
            現在のログ設定
        """
        return self._config

    def _print(self, message: str, file: TextIO | None = None) -> None:
        """メッセージを出力する

        Args:
            message: 出力するメッセージ
            file: 出力先（Noneの場合は標準出力）
        """
        if file is None:
            file = sys.stdout
        print(message, file=file)

    def _log_to_file(self, level: str, message: str) -> None:
        """ファイルにログ出力する

        Args:
            level: ログレベル文字列
            message: 出力するメッセージ
        """
        if self._log_file:
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            clean_message = self._strip_ansi(message)
            self._log_file.write(f"[{timestamp}] {level}: {clean_message}\n")
            self._log_file.flush()

    def _strip_ansi(self, text: str) -> str:
        """ANSIエスケープシーケンスを除去する

        Args:
            text: 処理対象のテキスト

        Returns:
            ANSIエスケープシーケンスを除去したテキスト
        """
        return self._ANSI_ESCAPE_PATTERN.sub("", text)

    def info(self, message: str) -> None:
        """情報メッセージを出力する（NORMAL以上）

        Args:
            message: 出力するメッセージ
        """
        if self._config.verbose_level >= VerboseLevel.NORMAL:
            self._print(message)
        self._log_to_file("INFO", message)

    def verbose(self, message: str) -> None:
        """詳細メッセージを出力する（VERBOSE以上）

        Args:
            message: 出力するメッセージ
        """
        if self._config.verbose_level >= VerboseLevel.VERBOSE:
            self._print(message)
        self._log_to_file("VERBOSE", message)

    def debug(self, message: str) -> None:
        """デバッグメッセージを出力する（DEBUG以上）

        Args:
            message: 出力するメッセージ
        """
        if self._config.verbose_level >= VerboseLevel.DEBUG:
            self._print(message)
        self._log_to_file("DEBUG", message)

    def error(self, message: str) -> None:
        """エラーメッセージを出力する（常に出力）

        Args:
            message: 出力するメッセージ
        """
        self._print(f"エラー: {message}", file=sys.stderr)
        self._log_to_file("ERROR", message)

    def warning(self, message: str) -> None:
        """警告メッセージを出力する（QUIET以上）

        Args:
            message: 出力するメッセージ
        """
        if self._config.verbose_level > VerboseLevel.QUIET:
            self._print(f"警告: {message}")
        self._log_to_file("WARNING", message)

    def create_progress(self) -> ProgressDisplay:
        """進捗表示インスタンスを作成する

        Returns:
            進捗表示インスタンス
        """
        return ConsoleProgressDisplay(
            use_color=self._config.use_color,
            use_emoji=self._config.use_emoji,
        )

    def log_command(self, command: list[str], output: str) -> None:
        """外部コマンド実行をログする（DEBUG以上）

        Args:
            command: 実行したコマンドとその引数
            output: コマンドの出力
        """
        cmd_str = " ".join(command)
        self.debug(f"実行: {cmd_str}")
        if output:
            for line in output.splitlines():
                self.debug(f"  > {line}")

    def log_conversion(self, source: Path, dest: Path, status: str) -> None:
        """ファイル変換をログする（VERBOSE以上）

        Args:
            source: 変換元ファイルパス
            dest: 変換先ファイルパス
            status: 変換ステータス
        """
        self.verbose(f"変換: {source.name} -> {dest.name} [{status}]")

    def log_summary(self, statistics: dict[str, Any]) -> None:
        """ビルドサマリを出力する（NORMAL以上）

        Args:
            statistics: ビルド統計情報
        """
        emoji = "\u2705" if self._config.use_emoji else "[OK]"
        self.info(f"{emoji} Build complete!")
        if "output_path" in statistics:
            size_mb = statistics.get("output_size", 0) / (1024 * 1024)
            self.info(f"   Output: {statistics['output_path']} ({size_mb:.1f} MB)")
        if "package_name" in statistics:
            self.info(f"   Package: {statistics['package_name']}")


class ConsoleProgressDisplay:
    """コンソール進捗表示

    ビルドパイプラインの各フェーズの進捗をコンソールに表示するクラス。
    進捗バーと絵文字を使用して視覚的にわかりやすい表示を行う。
    """

    # TYPE_CHECKINGブロック外でPipelinePhaseをインポートする必要があるため遅延インポート
    PHASE_EMOJI: dict[str, str] = {
        "analyze": "\U0001f50d",
        "extract": "\U0001f4e6",
        "convert": "\U0001f504",
        "build": "\U0001f528",
        "sign": "\U0001f50f",
    }

    PHASE_NAME: dict[str, str] = {
        "analyze": "Analyzing game structure",
        "extract": "Extracting assets",
        "convert": "Converting assets",
        "build": "Building APK",
        "sign": "Signing APK",
    }

    def __init__(self, use_color: bool = True, use_emoji: bool = True) -> None:
        """進捗表示を初期化する

        Args:
            use_color: カラー出力を使用するか
            use_emoji: 絵文字を使用するか
        """
        self._use_color = use_color
        self._use_emoji = use_emoji
        self._phase: PipelinePhase | None = None
        self._total = 0
        self._current = 0

    def start(self, phase: PipelinePhase, total: int) -> None:
        """フェーズ開始を表示する

        Args:
            phase: 開始するパイプラインフェーズ
            total: 処理対象の総数
        """
        self._phase = phase
        self._total = total
        self._current = 0
        emoji = self.PHASE_EMOJI.get(phase.value, "") if self._use_emoji else ""
        name = self.PHASE_NAME.get(phase.value, str(phase))
        prefix = f"{emoji} " if emoji else ""
        print(f"{prefix}{name}...")

    def update(self, current: int, message: str = "") -> None:
        """進捗を更新する

        Args:
            current: 現在の進捗（処理済みアイテム数）
            message: 追加の進捗メッセージ（オプション）
        """
        self._current = current
        if self._total > 0:
            percent = int((current / self._total) * 100)
            bar_width = 40
            filled = int(bar_width * current / self._total)
            bar = "\u2588" * filled + "\u2591" * (bar_width - filled)
            msg_part = f" {message}" if message else ""
            print(f"\r   [{bar}] {percent}%{msg_part}", end="", flush=True)

    def finish(self, success: bool, message: str = "") -> None:
        """フェーズ終了を表示する

        Args:
            success: フェーズが成功したか
            message: 終了メッセージ（オプション）
        """
        bar_width = 40
        full_bar = "\u2588" * bar_width
        if success:
            mark = "\u2713" if self._use_emoji else "done"
            print(f"\r   [{full_bar}] 100% {mark}")
        else:
            mark = "\u2717" if self._use_emoji else "failed"
            msg_part = f": {message}" if message else ""
            print(f"\r   [{full_bar}] {mark}{msg_part}")

File: internal/logger/logger.go
// Package logger は進捗表示およびログ出力を提供する。
//
// VerboseLevelに応じた出力制御を行い、CLIでのビルド進捗をユーザーにわかりやすく
// 表示するために使用される。
package logger

import (
	"fmt"
	"io"
	"regexp"
	"slices"
	"strings"
	"sync"
	"time"

	"github.com/na2na-p/mnemonic/internal/pipeline"
)

// VerboseLevel は詳細ログレベルを表す。
//
//   - Quiet: エラーのみ出力
//   - Normal: 進捗バーとサマリ出力
//   - Verbose: 変換ファイル一覧も出力（-vオプション）
//   - Debug: 外部コマンド実行ログも出力（-vvオプション）
type VerboseLevel int

// VerboseLevelの各段階。値の大小関係で出力するログの詳細度を比較する。
const (
	Quiet   VerboseLevel = -1
	Normal  VerboseLevel = 0
	Verbose VerboseLevel = 1
	Debug   VerboseLevel = 2
)

// ProgressDisplay はビルドパイプラインの各フェーズの進捗を表示するインターフェース。
type ProgressDisplay interface {
	// Start はフェーズ開始を表示する。
	Start(phase pipeline.Phase, total int)
	// Update は進捗を更新する。messageは追加の進捗メッセージ（不要な場合は空文字列）。
	Update(current int, message string)
	// Finish はフェーズ終了を表示する。messageは終了メッセージ（不要な場合は空文字列）。
	Finish(success bool, message string)
}

var ansiEscapePattern = regexp.MustCompile(`\x1b\[[0-9;]*m`)

// secretFlags は値が秘密情報になる引数名。apksignerの--ks-pass / --key-passと、
// keytoolの-storepass / -keypass。
var secretFlags = []string{"--ks-pass", "--key-pass", "-storepass", "-keypass"}

// BuildLogger はコンソールと注入されたログ出力先へビルドログを書き込む。
type BuildLogger struct {
	level  VerboseLevel
	stdout io.Writer
	stderr io.Writer
	file   io.Writer

	mu  sync.Mutex
	err error
}

// New は出力先を注入してBuildLoggerを生成する。fileがnilの場合はファイル出力を行わない。
//
// why not: ログファイルのパスを受け取ってここで開く形にはしない。追記するか切り詰めるかの
// 選択と書き込み完了後のCloseは、ファイルを開いた側が一貫して責任を持つ。
func New(level VerboseLevel, stdout, stderr, file io.Writer) *BuildLogger {
	return &BuildLogger{level: level, stdout: stdout, stderr: stderr, file: file}
}

func (l *BuildLogger) write(writer io.Writer, message string) {
	if _, err := fmt.Fprintln(writer, message); err != nil && l.err == nil {
		l.err = fmt.Errorf("ログの書き込みに失敗しました: %w", err)
	}
}

func (l *BuildLogger) log(level, consoleMessage, fileMessage string, console io.Writer, enabled bool) {
	l.mu.Lock()
	defer l.mu.Unlock()

	if enabled {
		l.write(console, consoleMessage)
	}
	if l.file != nil {
		clean := ansiEscapePattern.ReplaceAllString(fileMessage, "")
		l.write(l.file, prefixLines(fmt.Sprintf("[%s] %s: ", time.Now().Format("2006-01-02 15:04:05"), level), clean))
	}
}

// prefixLines はmessageの各行の先頭にprefixを付けて改行で連結する。messageの末尾の
// 改行は1つだけ行の終端として扱い、改行で区切った各行の末尾からCRを1つ除く。
//
// why not: 2行目以降を接頭辞無しで続けて書かない。grep ERRORのような行単位の抽出で
// 続きの行が落ち、別のエントリにも見える。
func prefixLines(prefix, message string) string {
	lines := make([]string, 0, strings.Count(message, "\n")+1)
	for line := range strings.SplitSeq(strings.TrimSuffix(message, "\n"), "\n") {
		lines = append(lines, prefix+strings.TrimSuffix(line, "\r"))
	}

	return strings.Join(lines, "\n")
}

// Info は情報メッセージを出力する（Normal以上）。
func (l *BuildLogger) Info(message string) {
	l.log("INFO", message, message, l.stdout, l.level >= Normal)
}

// Verbose は詳細メッセージを出力する（Verbose以上）。
func (l *BuildLogger) Verbose(message string) {
	l.log("VERBOSE", message, message, l.stdout, l.level >= Verbose)
}

// Debug はデバッグメッセージを出力する（Debug以上）。
func (l *BuildLogger) Debug(message string) {
	l.log("DEBUG", message, message, l.stdout, l.level >= Debug)
}

// Warning は警告メッセージを出力する（Quietより上のレベル）。
func (l *BuildLogger) Warning(message string) {
	l.log("WARNING", fmt.Sprintf("警告: %s", message), message, l.stdout, l.level > Quiet)
}

// Error はエラーメッセージを常に標準エラー出力へ出力する。
func (l *BuildLogger) Error(message string) {
	l.log("ERROR", fmt.Sprintf("エラー: %s", message), message, l.stderr, true)
}

// LogConversion はファイル変換の結果を記録する（Verbose以上）。
//
// why not: filepath.Baseで短くしない。bg/a.pngとfg/a.pngのように同名のファイルを
// 区別できなくなるため、呼び出し側が渡したパスをそのまま出す。
func (l *BuildLogger) LogConversion(source, dest, status string) {
	l.Verbose(fmt.Sprintf("変換: %s -> %s [%s]", source, dest, status))
}

// redactCommand はsecretFlagsの値を伏字にしたargvの写しを返す。
//
// why not: argvを書き換えない。呼び出し側は同じスライスをコマンドの実行にも使う。
func redactCommand(argv []string) []string {
	redacted := slices.Clone(argv)
	for i, arg := range redacted {
		for _, flag := range secretFlags {
			if arg == flag {
				if i+1 < len(redacted) {
					redacted[i+1] = "***"
				}
				break
			}
			prefix := flag + "="
			if strings.HasPrefix(arg, prefix) {
				redacted[i] = prefix + "***"
				break
			}
		}
	}

	return redacted
}

// LogCommand は外部コマンドの引数と成否を記録する（Debug以上）。
//
// why not: コマンドの出力とエラー値は受け取らない。署名ツールの出力や環境由来の
// 文字列は秘密を含み得るため、伏字化した引数と成否だけを記録する。
func (l *BuildLogger) LogCommand(argv []string, succeeded bool) {
	status := "失敗"
	if succeeded {
		status = "成功"
	}
	l.Debug(fmt.Sprintf("実行: %s [%s]", strings.Join(redactCommand(argv), " "), status))
}

// Err は最初に発生した書き込みエラーを返す。失敗がなければnil。
//
// why not: 書き込みに失敗しても以降の書き込みは止めない。ログファイルの障害で
// コンソール出力まで失わないよう、エラーは保持して扱いを呼び出し側に委ねる。
func (l *BuildLogger) Err() error {
	l.mu.Lock()
	defer l.mu.Unlock()

	return l.err
}
